apps/src/pythonlab/patches/pythonScriptUtils.ts

```diff
@@ -54,59 +61,162 @@ export function writeSource(
+// TODO: determine if we want to do the following:
+// - look for deleted folders or files.
+// - send an error message if a non-csv/txt file is updated (this is currently ignored).
+// - process hidden folders (we currently ignore them).
+// Tracked in https://codedotorg.atlassian.net/browse/CT-509
+export function getUpdatedSourceAndDeleteFiles(
+  source: MultiFileSource,
+  id: string,
   pyodide: PyodideInterface,
-  source: MultiFileSource
+  sendMessage: (message: PyodideMessage) => void
 ) {
-  Object.values(source.files).forEach(file => {
-    const filePath = getFilePath(file.id, source);
-    try {
-      pyodide.FS.unlink(filePath);
-    } catch (e) {
-      // TODO: log error better. We catch this because it should not prevent
-      // future runs.
-      console.warn(`error unlinking Pyodide file ${filePath}, ${e}`);
+  const workingDir = pyodide.FS.cwd();
+  const directoryData = pyodide.FS.lookupPath(workingDir, {}).node;
+  const directoryContents = Object.values(
+    directoryData.contents
+  ) as PyodidePathContent[];
+  const newSource = {...source};
+  updateAndDeleteSourceWithContents(
+    directoryContents,
+    newSource,
+    workingDir + '/',
+    DEFAULT_FOLDER_ID,
+    id,
+    pyodide,
+    sendMessage
+  );
+  return newSource;
+}
+
+function updateAndDeleteSourceWithContents(
+  contents: PyodidePathContent[],
+  source: MultiFileSource,
+  currentPath: string,
+  folderId: string,
+  id: string,
+  pyodide: PyodideInterface,
+  sendMessage: (message: PyodideMessage) => void
+) {
+  contents.forEach(content => {
+    const fileExtension = content.name.split('.').pop();
+    const fullPath = currentPath + content.name;
+    if (pyodide.FS.isFile(content.mode)) {
+      if (fileExtension === 'csv' || fileExtension === 'txt') {
+        const file = Object.values(source.files).find(
+          f => f.name === content.name
+        );
+        try {
+          const newContents = pyodide.FS.readFile(fullPath, {
+            encoding: 'utf8',
+          });
+          if (!file) {
+            const newFileId = getNextFileId(Object.values(source.files));
+            source.files[newFileId] = {
+              id: newFileId,
+              folderId,
+              name: content.name,
+              language: fileExtension,
+              contents: newContents,
+            };
+          } else {
+            file.contents = newContents;
+          }
+        } catch (e) {
+          console.warn(`could not read file ${content.name}, ${e}`);
+        }
+      } else {
+        if (!Object.values(source.files).some(f => f.name === content.name)) {
+          sendMessage({
+            type: 'error',
+            message: `You cannot write a ${fileExtension} file in Python Lab`,
+            id: id,
+          });
+        }
+      }
+      // Delete the file now that we have handled it.
+      pyodide.FS.unlink(fullPath);
+      // Do not create or iterate through folders that start with '.' as these are hidden folders.
```

Comment:
Do we need/want to skip hidden files? They'll have the same prefix, right?

Reply:
Late breaking update that we will show hidden folders and files. I'll add that to the follow up ticket.